### PR TITLE
Concatenate improvements

### LINKED ISSUE #1398: Better concatenation for binned data
The current implementation of `concatenate` is preliminary and has a number of shortcomings:

- No in-place operation is supported.
- When concatenating the inner dimension (or rather the dimension the bin data is ordered in) much faster implementations could be provided:
  - Just change indices of bin data is already contiguous in memory (no gaps between bins)
  - In-place move if bin data to remove gaps in the other cases.
- Slice based operation is probably not efficient for large dimension length, should in general switch to an implementation based on `transform` (probably in combination with `subspan_view` or similar).

Note that a lot of this may require extending the toolbox we have for binned data, which links into a couple of other issues. Make sure to discuss before making any implementation.

Update: It may make sense to just use `bin` internally to implement `concatenate`.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #1398 (still no in-place support or fast inner concat though).

- The implementation is now based on `bin` (or some of its internals), instead of a slow slice-by-slice copy. This appears to be faster, but it is not clear if it will ultimately yield the required performance.
- Dimensions with 2-D coords that are in conflict will now just lead to joining all bins along that dimension (see updated neutron powder diffraction tutorial).
- Fix index setup error in `element/bin.h` for cases of dropped events in multiple dims.

### CODE REVIEW

File: dataset/bin.cpp [hunk 1]
```diff
@@ -308,15 +351,86 @@ auto hide_masked(
...
+  builder.erase(dim);
+  TargetBins<T> target_bins(var);
+  builder.build(*target_bins, std::map<Dim, Variable>{});
+  auto [buffer, bin_sizes] = bin<DataArray>(var, *target_bins, builder.dims());
+  squeeze(bin_sizes, {dim});
+  const auto end = cumsum(bin_sizes);
+  const auto buffer_dim = buffer.dims().inner();
+  return make_bins(zip(end - bin_sizes, end), buffer_dim, std::move(buffer));
+}
+template Variable concat_bins<Variable>(const VariableConstView &, const Dim);
```
Comment (nvaytet) [inline dataset/bin.cpp:390]: Why are these needed?

File: docs/tutorials/neutron-diffraction.ipynb [hunk 2]
```diff
@@ -396,16 +396,12 @@
...
     "                    unit=sc.units.rad,\n",
     "                    values=np.linspace(0.0, np.pi/2, num=16))\n",
     "\n",
-    "# After conversion to d-spacing, bins of different spectra are not aligned.\n",
-    "# Therefore we first need to concatenate bins along d-spacing before concatenating spectra.\n",
-    "tmp = dspacing_sample.bins.concatenate('d-spacing')\n",
-    "focussed_sample = sc.groupby(tmp, 'theta', bins=theta).bins.concatenate('spectrum')\n",
-    "tmp = dspacing_vanadium.bins.concatenate('d-spacing')\n",
-    "focussed_vanadium = sc.groupby(tmp, 'theta', bins=theta).bins.concatenate('spectrum')\n",
+    "focussed_sample = sc.groupby(dspacing_sample, 'theta', bins=theta).bins.concatenate('spectrum')\n",
```
Comment (nvaytet) [inline docs/tutorials/neutron-diffraction.ipynb:403]: :+1:

**Review by nvaytet — COMMENTED** [review]
We should be careful about saying that #1398 would be fixed, because as you mention this does not implement in-place support nor fast inner concat.

I guess it is only implementing "use `bin` internally to implement `concatenate`."

File: dataset/bin.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply dataset/bin.cpp:390]: Essentially `concatenate` uses the same mechanism as `bin`, with the added twist that it may erase dimensions. Earlier in this branch I had extended `bin` (lower down) to support this, but I found the code hard to read. Splitting into three different top-level functions made this more easily understandable.

**Comment by SimonHeybrock** [comment]
> We should be careful about saying that #1398 would be fixed, because as you mention this does not implement in-place support nor fast inner concat.
> 
> I guess it is only implementing "use `bin` internally to implement `concatenate`."

True, we can open a new issue for those. I think performance looks adequate for what we need for 0.5

File: dataset/bin.cpp (hunk 1 above)
Comment (nvaytet) [reply dataset/bin.cpp:390]: I guess my question was more why do you need the `Variable` and `DataArray` template specifications here

Comment (SimonHeybrock) [reply dataset/bin.cpp:390]: Those are *explicit template instantiations*. They allow putting the template declaration in the header, but the implementation in the source file.

**Comment by nvaytet** [comment]
> True, we can open a new issue for those. I think performance looks adequate for what we need for 0.5
If you open the issue, then I will approve ;-)

**Comment by SimonHeybrock** [comment]
#1437

**Review by nvaytet — APPROVED** [review]
